relations/src/r1cs/constraint_system.rs

```diff
@@ -334,19 +333,23 @@ impl<F: Field> ConstraintSystem<F> {
             }
             outlined_lc.compactify();
 
-            let mut should_promote = false;
+            let mut should_outline = false;
 
+            // Check if outlining is worthwhile.
             let this_used_times = used_times[index.0];
             let this_len = outlined_lc.len();
 
             if this_used_times * this_len > this_used_times + 2 + this_len {
```

Comment:
Can we add a comment explaining the cost. (Namely that we change the arity from `this_len` repeated `this_used_times` to a constraint for `this_len * 1 = {new variable}` and then all of the `this_used_times` entries only have one usage.
```suggestion
            // Cost with no outlining = `lc_len * number of usages`
            // Cost with outlining is one constraint for `(lc_len) * 1 = {new variable}` and 
            // using that single new variable in each of the prior usages.
            // This has total cost `number_of_usages + lc_len + 2`
            if this_used_times * this_len > this_used_times + 2 + this_len {
```